parlai/core/metrics.py

```diff
@@ -42,6 +46,86 @@
 re_punc = re.compile(r'[!"#$%&()*+,-./:;<=>?@\[\]\\^`{|}~_\']')
 
 
+class Metric(ABC):
+    """
+    Base class for storing metrics.
+
+    Subclasses should define .report().
+    """
+
+    @abstractmethod
+    def report(self) -> float:
+        pass
+
+    def __str__(self) -> str:
+        return f'{self.report():.4g}'
+
+    def __repr__(self) -> str:
+        return f'{self.__class__} ({self.report():.4g})'
+
+
+class SumMetric(Metric):
+    """
+    Class that keeps a running sum of some metric.
+    """
+
+    def __init__(self, sum_: Union[Number, torch.Tensor]):
+
+        # Convert sum to float
+        if isinstance(sum_, torch.Tensor):
+            sum_ = sum_.item()
+        assert isinstance(sum_, Number)
+        sum_ = float(sum_)
+
+        self.sum = sum_
+
+    def __add__(self, other: 'SumMetric') -> 'SumMetric':
+        # NOTE: hinting can be cleaned up with "from __future__ import annotations" when
+        # we drop Python 3.6
```

Comment:
oh? i'm real curious about this, what's this?

Reply:
That's from https://stackoverflow.com/questions/41135033/type-hinting-within-a-class (see the second answer). In Python <= 3.6, you have to pass the class name as a string, but Python >= 3.7 allows for using the class name itself as the hinted type if you add `from __future__ import annotations`